tensorboard/notebook.py

```diff
@@ -278,9 +278,36 @@ def _display(port=None, height=None, print_message=False, display_handle=None):
 
 
 def _display_colab(port, height, display_handle):
-  # TODO(@wchargin): Implement this after merging this code into
-  # google3, where it's easier to develop and test against Colab.
-  raise NotImplementedError()
+  import IPython.display
+  shell = """
+    <script>
+      document.querySelector("base").href = "https://localhost:%s";
+      function fixUpTensorboard() {
+        const tftb = document.querySelector("tf-tensorboard");
+        tftb.removeAttribute("use-hash");
+      }
+      function rehydrate() {
+        for (const script of document.querySelectorAll("script")) {
+          const newScript = document.createElement("script");
+          newScript.type = script.type;
+          newScript.textContent = script.textContent;
+          document.body.appendChild(newScript);
+          newScript.remove();
+        }
+      }
+      fetch(".")
+        .then((x) => x.text())
+        .then((html) => void (document.body.innerHTML = html))
```

Comment:
Doesn't writing to the entire body end up overwriting the spacer div (and this script itself)?  is that desirable, versus writing the content into a narrower innerHTML target?

Reply:
> Doesn't writing to the entire body end up overwriting the spacer div
> (and this script itself)?

It does, yes, but by that time the output frame has already been sized
appropriately by the spacer, so it gets the job done. Overwriting the
script seems good to me; ideally, the DOM should be as close to the
actual TensorBoard `index.html` as feasible.

> is that desirable, versus writing the content into a narrower
> innerHTML target?

Eh… embedding the whole page within a narrower target (e.g., using an
actual `container` div rather than just a `spacer`) seems fishy to me.
That would mean that the DOM structure of TensorBoard proper is
structurally different from that of TensorBoard-in-Colab. I can see ways
in which this could be surprising/bad, and I don’t really see an
advantage.

I kind of view this script as `execve(2)`ing the page, which seems
reasonable enough to me. I don’t feel super strongly about this, though;
if you prefer, I can look into rendering everything inside a `<div>`.

Reply:
Fair enough, that's fine as long as the spacer still works reliably.

Reply:
I guess calling `google.colab.output.setIframeHeight` instead gets the
best of both worlds: no weird spacer, no weird DOM. If you don’t object
I’ll change the PR to use that form.